src/swarm/neo/request/RequestEventDispatcher.d

```diff
@@ -0,0 +1,767 @@
+
+    Request event dispatcher for use with multi-fiber requests.
+
+    Three types of events are handled:
+        1. "Signals" (ubyte codes) sent between fibers.
+        2. Messages received over the connection.
+        3. Connection readiness to send a payload.
+
+    The types of events are handled somewhat differently, as follows:
+        1. Only one request handler fiber may register to be notified of each
+           type of signal.
+        2. Messages reveived over the connection are expected to begin with a
+           byte representing the message type. Only one request handler fiber
+           may register to be notified of each type of message.
+        3. Each request handler fiber may only register once (at a time) for
+           notification of connection send readiness. Send readiness
+           notifications are handled in order of registration (i.e. fibers will
+           be notified of successive send readiness in the order in which they
+           registered to receive this notification.)
+
+    Copyright: Copyright (c) 2017 sociomantic labs GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+
+*******************************************************************************/
+
+module swarm.neo.request.RequestEventDispatcher;
+
+import ocean.transition;
+import ocean.core.SmartUnion;
+import ocean.util.log.Log;
+
+/*******************************************************************************
+
+    Static module logger.
+
+*******************************************************************************/
+
+static private Logger log;
+static this ( )
+{
+    log = Log.lookup("swarm.neo.request.RequestEventDispatcher");
+}
+
+
+/*******************************************************************************
+
+    Smart union of events which can be registered with the dispatcher.
+
+*******************************************************************************/
+
+public alias SmartUnion!(EventInUnion) EventIn;
+
+/*******************************************************************************
+
+    Union of events which can be registered with the dispatcher.
+
+*******************************************************************************/
+
+private union EventInUnion
+{
+    Message message;
+    Signal signal;
+    Send send;
+
+    istring toString ( )
+    {
+        // FIXME: ocean #106
+        return "dummy"; // Required in unittests builds for error formatting
+    }
+}
+
+/*******************************************************************************
+
+    Struct representing a fiber awaiting a certain message over the connection.
+    Each message is required to begin with a single byte representing the
+    message type.
+
+*******************************************************************************/
+
+public struct Message
+{
+    /// The type of the message being awaited.
+    ubyte type;
+}
```

Comment:
What is the reason for single-member structs like this -- why not a `ubyte message_type` field in `EventInUnion`?